cmd/manager/main.go

```diff
@@ -97,9 +104,11 @@ func main() {
 		caCertDir                 string
 		globalPullSecret          string
 	)
-	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&metricsAddr, "metrics-bind-address", "0", "The address for the metrics endpoint. Use :8443 for HTTPS or set to 0 to disable. Requires both tls-cert and tls-key.")
```

Comment:
I see in the Options struct that setting to 0 disables the entire metric server, so I answered that question for myself. In which case, that should be more directly stated in the help message here, i.e. "or set to 0 to disable the metrics server."

Was a decision made that we now want to have the metrics server disabled by default?

Reply:
We can improve the description if you see that is required, but renaming the flag is not an option because it is already in use, and changing it would result in a breaking change for us. So, I think we do not discuss it. 

The flag is intended to specify the port used for the metrics server. If set to 0, it will disable the metrics server.
See the RFC: https://docs.google.com/document/d/1hYOtHWcfoX_zWvy2YeolQBGuDgmP72vXMlrj9q9kPBQ/edit?tab=t.0#heading=h.x3tfh25grvnv

![Screenshot 2024-12-18 at 17 59 29](https://github.com/user-attachments/assets/605ee12f-57aa-45a5-936e-32be3c34a04c)

You can find there the link for the code in controller-runtime.

Reply:
Also, you don't have to use 8443 to serve HTTPS, you can use any port you want. Is this something we are wanting to enforce on users?

If we do want to force using 8443 or 0 to disable, then this flag should be rewritten as a boolean flag.

If we don't care which port users use for the metrics server, then this needs to have rewritten help text and probably a different default value.

Do we want to allow users to run the metrics server on plain HTTP?

As the code is right now, as a user I might read these help messages for the available flags and think I just need to provide a TLS cert and key to have my secure metrics server, but since metrics-bind-address defaults to 0, it's going to be disabled. 

I think depending on the answers to my questions above, we need to better enforce that the three flag options of keyfile, certfile, and metricsBindAddr are either all used together, or none are used--if a user uses one or two but not the third, then we should probably fail and tell the user why. Right now some of those edge cases of using only some of the options are not covered.

Reply:
> Also, you don't have to use 8443 to serve HTTPS, you can use any port you want. Is this something we are wanting to enforce on users?

We want to set 8443. See: 

https://github.com/operator-framework/operator-controller/pull/1475/files#diff-e40019a0cad999763c42de1c2a364c2fa74258f5d5cabc4c36306b50ddfac0cbR57-R60

> If we do want to force using 8443 or 0 to disable, then this flag should be rewritten as a boolean flag.

No, because tomorrow, we might want or need to change it.
Changing the type from string to boolean can cause breaking changes. 
Today, it uses 8443 by default for the metrics, and we want to avoid addressing breaking changes.

Reply:
It's not using 8443 by default, you're defaulting it to 0. That link to the manager.yaml file shows that we set 8443 by default in that manifest, but the manifest defaults do not necessitate a hard-coded value in the go code--users can change the manifest all they want, the code needs to be able to handle those kinds of changes. The current code on main allows the user to set any port they want, that's the purpose of the flag.

If we want to preserve as much of the current user experience as possible, then I see two different options, depending on if we want to allow users to run metrics without HTTPS or not:

- If we don't want to allow HTTP, then you need to make the metrics-bind-address flag default to 8443, change the message to something like "The address for the metrics endpoint. Set to 0 to disable the metrics server." and make tls-cert and tls-key required flags.
- If we do want to allow HTTP, then make the metrics-bind-address flag default to 8443, change the message to something like "The address for the metrics endpoint. Will serve on HTTPS if tls-cert and tls-key are provided. Set to 0 to fully disable the metrics server." and add logic for handling setting up the metrics server options without TLS and with the provided port

But all of those cases should be able to handle any valid port number passed to the argument

Reply:
Another thought I had was that in the code on 'main,' the ability to disable the metrics server is not exposed or explained to users--being able to set the flag to 0 to disable the server is a detail you only see if you dig into the controller-runtime package, so it's not obvious to me that we wanted users to be able to disable the server?

Reply:
## Why can't `metrics-bind-address` default to any value other than `0`?

We want to ensure that the metrics server is **disabled by default** (`0` if no port is specified). This follows a **security-by-default** principle, prioritizing safety and minimizing unnecessary exposure.

## Why not make `tls-cert` and `tls-key` required flags?

The solution should be able to run **without exporting metrics**. If metrics are not required, we simply do not want the metrics server running in such cases, avoiding unnecessary configuration or resources.

## Why change the description?

We recommend using **port 8443** for exporting metrics, as it is our **default port**.

Using any other port might lead to potential issues:
- Conflicts with ports already in use (e.g., webhook ports).  
- Inconsistencies with our documentation and examples, which often assume metrics are exported on **port 8443**.  

While we don’t block the use of other ports (to remain flexible if a change is needed), there’s no strong reason to suggest using an arbitrary port. Consistency and clarity make 8443 the preferred option.

But if you have a better suggestion for the description that still adding, recommending 8443(our default) I am totally fine to change it out.